src/main/java/org/jboss/threads/JBossExecutors.java

```diff
@@ -337,12 +314,21 @@ static ClassLoader getContextClassLoader(final Thread thread) {
      * @return the old context class loader
      */
     static ClassLoader getAndSetContextClassLoader(final Thread thread, final ClassLoader newClassLoader) {
-        final ClassLoader currentClassLoader = (ClassLoader) unsafe.getObject(thread, contextClassLoaderOffs);
-        if (currentClassLoader != newClassLoader) {
-            // not using setContextClassLoader to save loading the current one again
-            unsafe.putObject(thread, contextClassLoaderOffs, newClassLoader);
+        final SecurityManager sm = System.getSecurityManager();
+        if (sm != null) {
+            return AccessController.doPrivileged((PrivilegedAction<ClassLoader>) () -> {
```

Comment:
this can be a problem..if happen in the hot path :(

Reply:
Yeah. I'm thinking maybe we should have an MR JAR overlay here which just calls the target method on 24+.

Reply:
The alternative would be a `VarHandle` and require `--add-opens java.base/java.lang=xxx` where `xxx` is `org.jboss.threads` or `ALL-UNNAMED`. Not a good alternative IMO.

Reply:
Can't simply strip out the use of `SecurityManager` yet?

Reply:
IIRC WildFly will support the security manager until it no longer supports Jakarta EE 10, which is still in the future AFAIK. That said, we could probably do something like have a Java 24+ MR-JAR layer that skips it, perhaps, and use Unsafe on the lower layers?

Reply:
FWIW, I had stripped all Hibernate libraries from any reference to the SecurityManager - but this applies only to next-gen major versions, so consumers have the option to use the older one.

Reply:
WildFly and EAP need to continue to support the security manager. Especially for Jakarta EE 10 as it's required per the spec. Ripping out privileged blocks could potentially require WildFly to fork projects or stick with older versions, neither of which is ideal.

That said, we can't control what the JDK does. If we reach a point where the JDK removes the API's, then we cannot support that version of the JVM. However, until then we prefer projects keep privileged blocks in tact.

Reply:
Thanks @jamezp for intervening! Yet we could have such a change in a minor (or even major?) and only update selected versions of future Wildfly for which this is fine? I'd do that over MR, also because all such optimisations being discussed here don't quite belong in the long-term-maintained branches.

Reply:
TBH it's really up to the project. We need to continue support for the security manager for various reasons. If a project WildFly/EAP consumes removes it and it's a Red Hat/JBoss project, there is a good chance you'll have to support multiple branches/releases :)

What I mean is, it really comes down to what the project wants to do. Supporting two different release streams is fine, if the project is okay with that.

For the projects I lead, I just leave the security manger code because I personally don't want to have to support multiple release streams for who knows how long :)

Reply:
My easy fix for this is to use `Unsafe` up to JDK 23, and then use the normal accessors for 24+, since in 24+ there are no security manager checks and thus the abstraction does not do anything special anymore.